File: Array/subarray/maximum_subarray.kt
package Array.subarray

//https://leetcode.com/problems/maximum-subarray/
// Can refer to the notes
/*
Find subarray with biggest sum
 */
/*
TC: O(n)        SC: O(1)
2 integers here
currentSubarray
    will keep the running count of the current subarray we are focusing on.
maxSubarray
    will be our final return value. Continuously update it whenever we find a bigger subarray.

During iteration :
 1. Start with the 2nd elem, use the 1st one to init our variable
     If currentSubarray becomes negative, we know it isn't worth keeping, so throw it away.

 2. Remember to update maxSubarray every time we find a new maximum
    and return maxSubarray
 */
fun maxSubArray(nums: IntArray): Int {
    // Initialize our variables using the first element.
    var currentSubarray = nums[0]
    var maxSubarray = nums[0]

    // Start with the 2nd element since we already used the first one
    // to init our variable
    for (i in 1 until nums.size) {
        val num = nums[i]

        // If you run into negative how to skip?
        // If current_subarray is negative, throw it away. Otherwise, keep adding to it.
        currentSubarray = Math.max(num, currentSubarray + num)
        maxSubarray = Math.max(maxSubarray, currentSubarray)
    }
    return maxSubarray
}

File: String/String_Array/maximum_subarray.py
"""

Greedy like algorithm behind it here

nums = [-2, 1, -3, 4, -1, 2, 1, -5, 4]

cur_subarray
max_subarray
i = 0
-2 not worth keeping, so skip it here

i = 1
cur_subarray = 1
max_subarray = 1

i = 2
cur_subarray = -2
max_subarray = 1

i = 3
-2 (cur_subarray) prior not worth keeping here
cur_subarray = 2
max_subarray = 4

and the above is what we expected from this equation so far here and that's very important so far here



"""
from String.String_Array.findMode import List


class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        # Initialize our variables using the first element.
        current_subarray = max_subarray = nums[0]

        # Start with the 2nd element since we already used the first one.
        for num in nums[1:]:
            # If current_subarray is negative, throw it away. Otherwise, keep adding to it.
            current_subarray = max(num, current_subarray + num)
            max_subarray = max(max_subarray, current_subarray)

        return max_subarray